### DISCUSSION TITLE
USB passthrough /redirection

### DISCUSSION THREAD

**Discussion by kskarthik** [rustdesk/rustdesk#6014] [Feature Request]
Windows rdp has a mechanism for forwarding local USB devices to remote machines.

It appears that nomachine and splashtop both support the same feature.

This is necessary for using local USB devices and security keys in remote machines.

Since this is a feature request for all operating systems, as it does not exist in the software currently.
Include these links as resources

This feature can help support devices like drawing tablets
And maybe USB camera and microphones

References:

https://community.teamviewer.com/English/kb/articles/109554-security-key-redirection

https://www.splashtop.com/blog/leveraging-usb-device-redirection

https://superuser.com/questions/889973/no-machine-physical-device-forwarding#890245

**Comment by shamirahim111** [comment]
this is a much used feature that is sorely missing in rustdesk

**Reply by ruchit-bhatt** [reply]
+1000 for this

**Comment by ruchit-bhatt** [comment]
Any update on this ?

**Reply by shamirahim111** [reply]
@ruchit-bhatt no idea. maybe @rustdesk can help with this. dunno if this is even on their radar

**Comment by Pogimon3** [comment]
Don't think the team is taking this up right now. They are open to PRs, however.

Duplicate issue at: https://github.com/rustdesk/rustdesk/issues/2027

**Comment by ruchit-bhatt** [comment]
Any update on this? And @Pogimon3 this is completely different, it has nothing to do with usb drive mount remotely. Drawing tablet support will add support for pressure sensitivity for graphics related app.

**Comment by shodanx2** [comment]
For now use https://github.com/vadimgrn/usbip-win2
USB over IP once connected. 
You can run a script that make it work in one click 
But you have to port forward on your router, which is a bummer
Unless you use also maybe portable [softether](https://www.softether.org/) or [headscale](https://github.com/juanfont/headscale)

**Reply by ozFTW** [reply]
any similar solution for linux/macos?

**Comment by Hollywood25201** [comment]
Please add this ability. This means I can't use rustdesk with physical keys at all. love the software

**Comment by landside457** [comment]
Yea, passthrough for Smartcards (PIV based Cert Auth) would be great :)

**Comment by imanbarati** [comment]
is this thing done

**Comment by StealUrKill** [comment]
Question for those that are still subscribed to this.

**Are you wanting this on Windows only or for all operating systems?**

Asking because I have written and implemented a rust written **_USB Anywhere_** tool that does this. I currently use it via tailscale for my PIV and YubiKey. I can definitely make this work W11 to W11 within RustDesk. Getting the other operating systems to work might be a bit of a challenge.

### Issue
My biggest issue is this will not currently work with WebSocket enabled without having the server updated to accept a tunnel type connection on the pro side.

To put it simply, when WebSocket is turned on RustDesk chops all the traffic into small labeled messages. That is perfect for a normal remote session, but port forwarding needs a clean open pipe to push the USB data through, and the WebSocket path does not hand it one right now. The USB tool just tunnels its traffic over that port forwarding pipe, so RustDesk never has to understand USB at all, it only has to carry the bytes back and forth. Over a normal direct or relay connection it carries the USB traffic just fine.

This is not only a USB thing either, the regular TCP tunneling already built into RustDesk hits the same wall over WebSocket, so one small server change would actually implement two features at once. That change is a server side update to add a dedicated tunnel route for it. So a self hosted server could be patched to allow it, but the official hosted server would need that change first.

### Preview of USB Anywhere
<img width="969" height="1120" alt="image" src="https://github.com/user-attachments/assets/e1a899ed-ad1c-4363-a82d-b2efe16bb7ce" />

**Reply by Hollywood25201** [reply]
> Question for those that are still subscribed to this.
> 
> **Are you wanting this on Windows only or for all operating systems?**
> 
> Asking because I have written and implemented a rust written **_USB Anywhere_** tool that does this. I currently use it via tailscale for my PIV and YubiKey. I can definitely make this work W11 to W11 within RustDesk. Getting the other operating systems to work might be a bit of a challenge.
> ### Issue
> 
> My biggest issue is this will not currently work with WebSocket enabled without having the server updated to accept a tunnel type connection on the pro side.
> 
> To put it simply, when WebSocket is turned on RustDesk chops all the traffic into small labeled messages. That is perfect for a normal remote session, but port forwarding needs a clean open pipe to push the USB data through, and the WebSocket path does not hand it one right now. The USB tool just tunnels its traffic over that port forwarding pipe, so RustDesk never has to understand USB at all, it only has to carry the bytes back and forth. Over a normal direct or relay connection it carries the USB traffic just fine.
> 
> This is not only a USB thing either, the regular TCP tunneling already built into RustDesk hits the same wall over WebSocket, so one small server change would actually implement two features at once. That change is a server side update to add a dedicated tunnel route for it. So a self hosted server could be patched to allow it, but the official hosted server would need that change first.
> ### Preview of USB Anywhere
> <img alt="image" width="969" height="1120" src="https://private-user-images.githubusercontent.com/35749471/613921301-e1a899ed-ad1c-4363-a82d-b2efe16bb7ce.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODI1Mjg1ODYsIm5iZiI6MTc4MjUyODI4NiwicGF0aCI6Ii8zNTc0OTQ3MS82MTM5MjEzMDEtZTFhODk5ZWQtYWQxYy00MzYzLWE4MmQtYjJlZmUxNmJiN2NlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjI3VDAyNDQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwMDRhOTFhMTA1MzFjNTQ0YTA4ZjAzZTM1ZTU4NzEzZDM1NGRjZjdmMzY4ZjY4NWYzYTI2MjcyZGVkMTRjNjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.ZESRBrWyfmM6bbbIdNpOK_TFsyRg26TcaFs-AENhmMw">

This looks amazing!! I would definitely love a version that works with Linux, I personally don't have a need for Mac, but I'm sure someone would.

Thanks for putting this out for us!

**Reply by StealUrKill** [reply]
> > Question for those that are still subscribed to this.
> > **Are you wanting this on Windows only or for all operating systems?**
> > Asking because I have written and implemented a rust written **_USB Anywhere_** tool that does this. I currently use it via tailscale for my PIV and YubiKey. I can definitely make this work W11 to W11 within RustDesk. Getting the other operating systems to work might be a bit of a challenge.
> > ### Issue
> > My biggest issue is this will not currently work with WebSocket enabled without having the server updated to accept a tunnel type connection on the pro side.
> > To put it simply, when WebSocket is turned on RustDesk chops all the traffic into small labeled messages. That is perfect for a normal remote session, but port forwarding needs a clean open pipe to push the USB data through, and the WebSocket path does not hand it one right now. The USB tool just tunnels its traffic over that port forwarding pipe, so RustDesk never has to understand USB at all, it only has to carry the bytes back and forth. Over a normal direct or relay connection it carries the USB traffic just fine.
> > This is not only a USB thing either, the regular TCP tunneling already built into RustDesk hits the same wall over WebSocket, so one small server change would actually implement two features at once. That change is a server side update to add a dedicated tunnel route for it. So a self hosted server could be patched to allow it, but the official hosted server would need that change first.
> > ### Preview of USB Anywhere
> > <img alt="image" width="969" height="1120" src="https://private-user-images.githubusercontent.com/35749471/613921301-e1a899ed-ad1c-4363-a82d-b2efe16bb7ce.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODI1Mjg1ODYsIm5iZiI6MTc4MjUyODI4NiwicGF0aCI6Ii8zNTc0OTQ3MS82MTM5MjEzMDEtZTFhODk5ZWQtYWQxYy00MzYzLWE4MmQtYjJlZmUxNmJiN2NlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjI3VDAyNDQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwMDRhOTFhMTA1MzFjNTQ0YTA4ZjAzZTM1ZTU4NzEzZDM1NGRjZjdmMzY4ZjY4NWYzYTI2MjcyZGVkMTRjNjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.ZESRBrWyfmM6bbbIdNpOK_TFsyRg26TcaFs-AENhmMw">
> 
> This looks amazing!! I would definitely love a version that works with Linux, I personally don't have a need for Mac, but I'm sure someone would.
> 
> Thanks for putting this out for us!

FYI, It is not out. I was just saying it can work with existing codebase on non WebSocket clients. If there is a path forward with the TCP Tunneling for pro users with WebSocket, this could be implemented as well. Linux support is doable as my my work mostly came from a Linux C++ repo that I ported and adapted to Rust for Windows. I can probably add the Windows support fairly quickly but only having Windows without WebSocket support is probably something the owner would not want.

Just like I have a clipboard clear on disconnect for Windows or Resizing screen based on window content for Windows. Without universal support, these would most likely be a no-op unless someone else would work on it with me.

**Reply by SedonD** [reply]
> Are you wanting this on Windows only or for all operating systems?

One more vote for Linux

**Comment by rustdesk** [comment]
@StealUrKill this rustdesk clone https://sourcecraft.dev/rudesktop/rudesktop-client has implemented usb redirection. It will be great if someone can migrate it into rustdesk. It is not hard with the help of claude/codex.

**Reply by StealUrKill** [reply]
> @StealUrKill this rustdesk clone https://sourcecraft.dev/rudesktop/rudesktop-client has implemented usb redirection. It will be great if anyone can migrate it into rustdesk. It is not hard with the help with claude/codex.

I can definitely get windows. Probably Linux. I have never touched a Mac and don't have one. But once I get a few more projects cleaned up I'll look into windows implementation and testing.

**Reply by StealUrKill** [reply]
> @StealUrKill this rustdesk clone https://sourcecraft.dev/rudesktop/rudesktop-client has implemented usb redirection. It will be great if someone can migrate it into rustdesk. It is not hard with the help of claude/codex.

I have been working on the USB over IP side of this, using the same usbip win2 client driver that RUDesktop includes. I wanted to flag an important issue for anyone adding USB device sharing to RustDesk.

Sharing a streaming USB device, such as a webcam, can blue screen the client PC.

A webcam sends a constant stream of data, so the client can have hundreds of USB requests waiting at once. If the webcam is unplugged on the sharing side while it is still streaming, every outstanding URB gets cancelled at the same time. The driver then sends one cancel message, CMD_UNLINK, for each URB. That burst of cancel messages is the storm that causes the crash.

The issue is in how the usbip win2 UDE driver handles that cleanup. It sends the burst of cancel messages from inside the Windows kernel. In one driver version, this can overload the network path and cause a DPC_WATCHDOG crash with bugcheck 0x133. In the newer driver, it can touch memory that was already freed and crash with bugcheck 0x7E or 0x1E. I have confirmed both from crash dumps.

Any app using this driver with a high bandwidth streaming device can run into it. A sudden network disconnect or app crash may cause the same problem because it forces the same cleanup path. Low bandwidth devices like dongles and keyboards do not seem to trigger it.

**Reply by StealUrKill** [reply]
> @StealUrKill this rustdesk clone https://sourcecraft.dev/rudesktop/rudesktop-client has implemented usb redirection. It will be great if someone can migrate it into rustdesk. It is not hard with the help of claude/codex.

Per AI

https://github.com/vadimgrn/usbip-win2/blob/74f5a7fa8a4991f61bcee3ec846d3b182184f05d/drivers/ude/device_ioctl.cpp#L541-L558

A few things make this the key spot:

It's the root cause. The ::send(...) on line 552 is the per-URB CMD_UNLINK network send, and both crash variants (the 0x133 DPC watchdog and the newer 0x7E/0x1E use-after-free) trace back to it.
It's where a fix would go — the problem and the solution are in the same view.
It's basically self-documenting. Line 548 already has if (get_flag(dev.unplugged)) { … do not send unlink } — the driver already skips this send when it knows the device is unplugged, so it already treats this send as unsafe during teardown. The fix looks like it's just extending that same skip to the mass-cancel/teardown path.
And the loop that fires it once per outstanding request — the "burst" that overwhelms the kernel — is here:

https://github.com/vadimgrn/usbip-win2/blob/74f5a7fa8a4991f61bcee3ec846d3b182184f05d/drivers/ude/device.cpp#L181-L182

**Comment by shodanx2** [comment]
Hello,
Are you guys aware that linux has usb over ip support out of the box.
And there is also a windows compatible version ?

windows version
https://github.com/dorssel/usbipd-win

and on linux it is part of the package "linux-tools-generic" and it comes with the kernel
https://github.com/torvalds/linux/tree/master/tools/usb/usbip

Rustdesk should be compatible with that, it has been the standard way of doing usb over network for a long time.
Don't use proprietary "VirtualHere" 

I don't know if "USB Anywhere" is the same as there proprietary versions

https://www.digi.com/resources/videos/anywhereusb-configuration-utility
https://github.com/eclypsium/USBAnywhere
https://anywhereusb.co.uk/anywhereusb-faqs/

But, rustdesk definitely should not become copyright-encumbered for this basic functionality !

And BTW, rustdesk doesn't need to support these, you can script it, but the server needs to be installed on the remote host and that means installing a driver at least once for virtual usb as far as I know.

**Reply by StealUrKill** [reply]
Yes that is where I got the idea from. lol

You need 2 different drivers to be installed on both ends. If remote client has device or if local client has device.

I was already going to use those sources as that is what I built off of.

<img width="569" height="287" alt="image" src="https://github.com/user-attachments/assets/b8056743-6e4c-4e56-b386-0f9b1d4fdec9" />

Only issue is the driver for usbipd-win uses Virtual box driver. I do not know what the licensing from that would change here. Maybe someone will know more...

<img width="1793" height="1361" alt="image" src="https://github.com/user-attachments/assets/7c1c34ad-1692-4a87-a88c-d65590e3e729" />

**Reply by shodanx2** [reply]
Great ! I was worried it was going to another proprietary UI like the VirtualHere situation !

Looking over the driver situation, I found someone else also made a GUI frontend for [usbipd-win](https://github.com/dorssel/usbipd-win)

It is mentionned in the following issue
https://github.com/dorssel/usbipd-win/issues/382
which became a show-and-tell discussion
https://github.com/dorssel/usbipd-win/discussions/389

From that I found a list of the other usbipd-win GUI 
https://github.com/dorssel/usbipd-win/wiki/Graphical-User-Interfaces

There is some discussion as to the license of the virtualbox usb over ip driver, I could not clarify. 
The author of usbipd-win says it is GPL2

Reading from this

https://docs.oracle.com/en/virtualization/virtualbox/6.0/user/intro-installing.html

The base virtualbox package includes OHCI USB 1.1 support and is GPL2.

But EHCI and xHCI are part of the virtual box extension package which is under the Virtual box PUEL license by the dystopian engineering company Oracle.

https://www.virtualbox.org/wiki/VirtualBox_PUEL

So, yeah, I hope those FIDO keys work with USB 1.1 !

**Reply by StealUrKill** [reply]
> Great ! I was worried it was going to another proprietary UI like the VirtualHere situation !
> 
> Looking over the driver situation, I found someone else also made a GUI frontend for [usbipd-win](https://github.com/dorssel/usbipd-win)
> 
> It is mentionned in the following issue [dorssel/usbipd-win#382](https://github.com/dorssel/usbipd-win/issues/382) which became a show-and-tell discussion [dorssel/usbipd-win#389](https://github.com/dorssel/usbipd-win/discussions/389)
> 
> From that I found a list of the other usbipd-win GUI https://github.com/dorssel/usbipd-win/wiki/Graphical-User-Interfaces
> 
> There is some discussion as to the license of the virtualbox usb over ip driver, I could not clarify. The author of usbipd-win says it is GPL2
> 
> Reading from this
> 
> https://docs.oracle.com/en/virtualization/virtualbox/6.0/user/intro-installing.html
> 
> The base virtualbox package includes OHCI USB 1.1 support and is GPL2.
> 
> But EHCI and xHCI are part of the virtual box extension package which is under the Virtual box PUEL license by the dystopian engineering company Oracle.
> 
> https://www.virtualbox.org/wiki/VirtualBox_PUEL
> 
> So, yeah, I hope those FIDO keys work with USB 1.1 !

I am not emulating a virtual controller for the guest. I am using VBoxUSBMon + VBoxUSB purely as a host-side capture/filter driver. It grabs the real physical device on the host at whatever speed it actually runs, and code forwards raw URBs over USB/IP. The speed ceiling is the host's physical USB stack, not VirtualBox.

**Comment by sleepyeb** [comment]
as a user with a bunch of rustdesk machines, yes i can setup my own system but wouldn't it be great if rust desk just did this out of the box for us. 

It would be a fantastic feature i am forever plugging in a usb then using the file transfer to copy a file onto the remote machine. I also use yubikey, man, having a pre-configured usb port that each time i connect to my workstation i can just auth with my key then auth on the workstation with that same key would be incredible.
